src/main/scala/io/ino/solrs/future/Future.scala

```diff
@@ -0,0 +1,88 @@
+package io.ino.solrs.future
+
+trait Future[+T] {
+  def map[S](f: T => S): Future[S]
+  def flatMap[S](f: T => Future[S]): Future[S]
+}
+
+trait Promise[T <: Any] {
+  def future: Future[T]
+  def success(value: T): Unit
+  def failure(exception: Throwable): Unit
+}
+
+trait Factory[RFT[_]] {
+  def newPromise[T]: Promise[T]
+
+  def toBase[T]: (Future[T] => RFT[T])
+}
+
+/* plain Scala default implementation */
+object ScalaFactory extends Factory[scala.concurrent.Future] {
+  import scala.util.{ Success, Failure }
+
+  def toBase[T] = (f: Future[T]) => {
+    f match {
+      case sf: ScalaFuture[T] =>
+        sf.inner
+      case _ => throw new Exception("Wrong future type")
+    }
+  }
+
+  class ScalaFuture[+T](f: scala.concurrent.Future[T])
+      extends Future[T] {
+    val inner = f
+    //please complete this before usage if needed
+    lazy val executionContextPromise = scala.concurrent.Promise[scala.concurrent.ExecutionContext]
```

Comment:
I don't get the following, how's that intended? How shall users be able to provide an EC?